### ISSUE TITLE
crypto_pwhash_str  : When defined, the output format must be a string

### ISSUE THREAD

**Issue by Ludea** [jedisct1/libsodium.js#120]
Hello,

I use libsodium.js into an express app.
I would like to store password into a database, with argon hash.

So I use crypto_pwhash_str function. 
But the output give an error: When defined, the output format must be a string

```
Here is my code :
const libsodium = require('libsodium-wrappers-sumo');
(async() => {
  await libsodium.ready;
  const sodium = libsodium;
var hashed_password = [sodium.crypto_pwhash_STRBYTES];
module.exports.password_hash = function(password, res){
if (sodium.crypto_pwhash_str
    (hashed_password, password , password.length,
     sodium.crypto_pwhash_OPSLIMIT_SENSITIVE, sodium.crypto_pwhash_MEMLIMIT_SENSITIVE) != 0) {
    /* out of memory */
}
    return res(hashed_password);
}
```

**Comment by Ludea** [comment]
I swith to get the string from the function, it still ive the same issues

```
module.exports.password_hash = function(password, res){
 var hashed_password = sodium.crypto_pwhash_str(password , password.length, sodium.crypto_pwhash_OPSLIMIT_INTERACTIVE, sodium.crypto_pwhash_MEMLIMIT_SENSITIVE);
    return res(hashed_password);
}
```

**Comment by jedisct1** [comment]
You don't need to provide the length either.

**Comment by Ludea** [comment]
ok, thanks for help ! 
Now I thinks it is a memory issue.
I have this message : `abort("Cannot enlarge memory arrays. Either (1) compile with  -s TOTAL_MEMORY=X  with X higher than the current value 83886080, (2) compile with  -s ALLOW_MEMORY_GROWTH=1  which allows increasing the size at runtime, or (3) if you want malloc to return NULL (0) instead of this abort, compile with  -s ABORTING_MALLOC=0 ") `

**Comment by jedisct1** [comment]
Is it with `pwhash_MEMLIMIT_INTERACTIVE`?

**Comment by Ludea** [comment]
it is working with pwhash_MEMLIMIT_INTERACTIVE !
I switch only OPSLIMIT_SENSITIVE .
Thanks for help :)

**Comment by Ludea** [comment]
Sorry for reopen it, but I can not read the dist/modules-sumo correctly.

I would like to use crypto_pwhash_str_verify function. 
But I need only 2 arguments, the length is not required?
And the function return an int or a value ?

**Comment by jedisct1** [comment]
The length is never required in the Javascript bindings.

The function returns `true` if the password appears to be correct for the given hash, and `false` otherwise.

**Comment by Ludea** [comment]
Ok, it works !
Thanks for your help

**Comment by jedisct1** [comment]
Awesome!

**Comment by ryancwalsh** [comment]
@jedisct1 I'm loving the idea of libsodium.js! :-)

I haven't been able to find many docs or examples of how to use it or a list of functions and the arguments they require.

I've been reading https://download.libsodium.org/doc/password_hashing/the_argon2i_function.html and lots of docs and examples like https://github.com/mdp/gibberish-libsodium/blob/master/index.js, but I still haven't figured out how to get that simple demo to work.

Here is a runnable and editable sandbox demo: https://codesandbox.io/s/9y8nqz9m1r

Basically, what I'm currently stuck on is:

```
    var key = sodium.crypto_pwhash(KEYLEN, password,  salt, OPSLIMIT, MEMLIMIT,
       1, //what are valid algorithm arguments?
      "text" //is this a valid outputFormat?
    );

    console.log("key", key); //why does this neither print nor throw an exception?
```

Thanks so much for your library. And thanks for any tips.

**Comment by ryancwalsh** [comment]
@jedisct1 I've made a little bit of progress I think: https://codesandbox.io/s/m7mx5yqzrx?autoresize=1&expanddevtools=1&hidenavigation=1

I now see an error that might be a clue: `abort("Cannot enlarge memory arrays. Either (1) compile with  -s TOTAL_MEMORY=X  with X higher than the current value 16777216, (2) compile with  -s ALLOW_MEMORY_GROWTH=1  which allows increasing the size at runtime, or (3) if you want malloc to return NULL (0) instead of this abort, compile with  -s ABORTING_MALLOC=0 ") at Error`

I also tried decreasing MEMLIMIT down continually until eventually it was 8191, and then the error changed to:

```
Error: invalid usage
    at v (https://m7mx5yqzrx.codesandbox.io/node_modules/libsodium-wrappers/dist/modules/libsodium-wrappers.js:1:2859)

    at Object.ke [as crypto_pwhash] (https://m7mx5yqzrx.codesandbox.io/node_modules/libsodium-wrappers/dist/modules/libsodium-wrappers.js:1:28381)

    at Object.exportsSelf.encrypt (https://m7mx5yqzrx.codesandbox.io/src/index.js:32:24)

    at eval (https://m7mx5yqzrx.codesandbox.io/src/index.js:67:31)

    at mightThrow (https://m7mx5yqzrx.codesandbox.io/node_modules/jquery/dist/jquery.js:3534:29)

    at process (https://m7mx5yqzrx.codesandbox.io/node_modules/jquery/dist/jquery.js:3602:12)
```

Then I figured... maybe sandbox machines are weird, and I should try my local machine again.

On my own local machine, if I use `sodium.crypto_pwhash_MEMLIMIT_MIN`, I get this error: `TypeError: "Decoding failed."` at the line that is trying to generate `key` from `sodium.crypto_pwhash()`.

Any ideas what I'm doing wrong?

Thanks!

**Comment by ryancwalsh** [comment]
I finally got it working: https://stackoverflow.com/a/51814273/470749
